src/verifast1.ml

```diff
@@ -3189,18 +3267,70 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
           static_error l (match language with CLang -> "No such function: " ^ g | Java -> "No such method or function: " ^ g) None
       in
       if language = CLang || classmap = [] then func_call () else
-      let try_qualified_call tn es args fb on_fail =
+      let try_qualified_call tn es args fb class_targs on_fail =
+        let class_tparams = match try_assoc tn classmap with
+            Some {ctpenv} -> ctpenv
+            | None -> let InterfaceInfo (_, _, _, _, _, tparams) = List.assoc tn interfmap in tparams
+        in
+        let class_targs_c = List.length class_targs in
+        let class_tparams_c = List.length class_tparams in 
+        if (class_targs_c <> class_tparams_c && inAnnotation = None )
+          then static_error l (Printf.sprintf 
+            "The amount of type arguments for %s is not conform with the amount of type parameters (%s) the class expects: %s. and binding: %s"
+            g
+            (String.concat ", " class_tparams)
+            (String.concat ", " (List.map string_of_type class_targs))
+            (if fb = Instance then "Instance" else "Static")) None
+        else 
         let ms = get_methods tn g in
         if ms = [] then on_fail () else
         let argtps = List.map (fun e -> let (_, tp, _) = (check e) in tp) args in
+        (* Select the real methods with the correct amount of type parameters *)
+        let ms = List.map (fun (sign, (tn', lm, gh, rt, xmap, pre, post, epost, terminates, fb', v, abstract, meth_tparams)) ->
+          let wtargs = if (List.length meth_tparams = List.length targes || gh = Ghost) then List.map (fun tp -> check_pure_type (pn,ilist) meth_tparams tp gh) targes else

```

Comment:
Dit lijkt verkeerd: de method type parameters zijn niet in scope in de method call type arguments?? Je kan targes eens en voor altijd checken buiten de context van een specifieke callee?